csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -755,588 +755,499 @@ module ControlFlow {
-        result = last(ac.getInitializer(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastArrayCreationLengthArgument(
-        ArrayCreation ac, int i, Completion c
+      private ControlFlowElement lastRecSpecific(
+        ControlFlowElement cfe, Completion c1, Completion c2
       ) {
-        not ac.isImplicitlySized() and
-        result = last(ac.getLengthArgument(i), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastIfStmtCondition(IfStmt is, Completion c) {
-        result = last(is.getCondition(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastIfStmtThen(IfStmt is, Completion c) {
-        result = last(is.getThen(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastIfStmtElse(IfStmt is, Completion c) {
-        result = last(is.getElse(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastSwitchExpr(Switch s, Completion c) {
-        result = last(s.getExpr(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastSwitchStmtStmt(SwitchStmt ss, int i, Completion c) {
-        result = last(ss.getStmt(i), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastSwitchStmtCaseStmt(SwitchStmt ss, int i, Completion c) {
-        result = last(ss.getStmt(i).(CaseStmt).getBody(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastCasePattern(Case case, Completion c) {
-        result = last(case.getPattern(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastCaseCondition(Case case, Completion c) {
-        result = last(case.getCondition(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastCaseBody(Case case, Completion c) {
-        result = last(case.getBody(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastLoopStmtCondition(LoopStmt ls, Completion c) {
-        result = last(ls.getCondition(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastLoopStmtBody(LoopStmt ls, Completion c) {
-        result = last(ls.getBody(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastForeachStmtIterableExpr(ForeachStmt fs, Completion c) {
-        result = last(fs.getIterableExpr(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastForeachStmtVariableDecl(ForeachStmt fs, Completion c) {
-        result = last(fs.getVariableDeclExpr(), c) or
-        result = last(fs.getVariableDeclTuple(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastJumpStmtChild(JumpStmt js, Completion c) {
-        result = last(js.getChild(0), c)
-      }
-
-      pragma[nomagic]
-      ControlFlowElement lastTryStmtFinally(TryStmt ts, Completion c) {
-        result = last(ts.getFinally(), c)
+        exists(TLastRecComputation rec |
+          result = lastRec(cfe, c1, rec) and
```

Comment:
Can we simplify this to `result = lastRec(cfe, c1, TLastRecSpecificCompletion(c2))`

Reply:
Yes.